src/lock/optiql.cpp

```diff
@@ -181,10 +198,30 @@ OptiQL::LockX()  //
 {
   auto qid = GetQID();
   auto *qnode = &(_qnodes[qid]);
+  const auto new_tail = (static_cast<uint64_t>(qid) << kQIDShift) | kXLock;
 
-  throw std::runtime_error{"not implemented yet"};
+  auto cur = lock_.load(kRelaxed);
+  while (true) {
+    qnode->ver.store(cur & kLockMask, kRelaxed);
+    if (lock_.compare_exchange_weak(cur, new_tail, kAcqRel, kRelaxed)) break;
+    CPP_UTILITY_SPINLOCK_HINT
+  }
+
+  auto pred_qid = (cur & kQIDMask) >> kQIDShift;
```

Comment:
```suggestion
  const auto pred_qid = (cur & kQIDMask) >> kQIDShift;
```

もしくは，この次の行でしか使わないので直に値を埋め込んでもよいです．変数名が `pred_node` な時点で意図はほぼ伝わりますし．

また， `pred_node` 自体もelseで分岐した先でしか使わないので，ここではなく分岐した後に作成しましょうか．

Reply:
いったん直に値を埋め込む形で調整してみます